### ISSUE TITLE
Please resolve conflicting statements about license. 

### ISSUE THREAD

**Issue by brandondrew** [gettalong/hexapdf#118]
There are several conflicting criteria given on the web site regarding when a paid license is required.   

> The AGPL puts some restrictions in place to make sure that the community benefits from changes to the code. For example, if you use HexaPDF in an application and distribute that application, you have to make the source code of the whole application available under the AGPL. The same applies if your application is used over a network.

> If you want to use HexaPDF in a commercial setting, you need to purchase a commercial license.

1. When there are changes made to the HexaPDF code.  
2. When the HexaPDF code is added to an app (such as when used as a gem—which does not involve changing HexaPDF at all) that is used over a network—therefore any Rails app ever.  
3. When it is used in a commercial setting.  

In my case I would use it in a Rails app for a university, probably not making any changes to the HexaPDF code.  So only criteria  number 2 would apply. 

I would never criticize your choice to require a paid license for your hard work: you have a right to earn money from it.  But if # 2 is sufficient to require a paid license, then why are 1 and 3 even mentioned?  Please clarify.  And if # 2 is not sufficient, can it be removed?

**Comment by gettalong** [comment]
Thanks for bringing this up! Could you clarify where you found the three points below the quoted text?

Generally, if you don't want to adhere to the AGPL or can't adhere to it (e.g. because you can't make the code that uses HexaPDF available under AGPL), then you need a commercial license.

So the easiest way to think about this is: Can I do what I need using the AGPL licensed version? If yes, then you don't need a commercial license.

In your case I guess that you probably need a commercial license but please contact someone with legal expertise.

**Comment by gettalong** [comment]
@brandondrew Coming back to this I realized that the three points you list are just an interpretation of the quoted text - sorry for the misunderstanding!

I have updated the text in question and I hope that it is clearer now. As I wrote before, the commercial license is needed if one can't adhere to the AGPL..

**Comment by brandondrew** [comment]
@gettalong 

> As I wrote before, the commercial license is needed if one can't adhere to the AGPL.

Thanks for replying, but since the question is aimed at trying to get a clear statement of _how I can know if I can adhere to the AGPL_ saying "the commercial license is needed if one can't adhere to the AGPL" just brings me back to the question again.

I can't think of _any_ application that is _neither_ distributed _nor_ used over a network.  So one interpretation of your statement is that _every application that uses this must itself be fully open source *or* must use a commercial license._  If that is your position, could you please clarify?  

Again, there is nothing wrong with requiring a commercial license.  But it would be appreciated if you could make clear exactly _when_ you require that.

Thanks.

**Comment by gettalong** [comment]
@brandondrew I'm sorry but I'm not a lawyer and I really don't know if "every application that uses this must itself be fully open source or must use a commercial license" is true. Generally I would say that if you don't want to invest time to determine whether you can use the AGPL, your statement is an easy way to think about which license to choose. However, there may be a middle ground somewhere depending on your use of HexaPDF.

If you use a software, you need to adhere its license. In case of HexaPDF you have a choice of two licenses and you can pick whichever suits your case best. You yourself have to determine which license is appropriate for your use-case. If you are developing an application in a professional environment (e.g. as employee of a company), you most probably also have access to a lawyer whom you can ask.

**Comment by brandondrew** [comment]
> I'm sorry but I'm not a lawyer and I really don't know

@gettalong  If _even you_ do not understand the license, despite having chosen it for your project, then surely you cannot expect anyone else who is not a lawyer to understand it.

> If you are developing an application in a professional environment (e.g. as employee of a company), you most probably also have access to a lawyer whom you can ask.

I hope you realize the effects of your choices, and I hope you realize that developers are not likely to even consider a project which forces them to consult a lawyer unless all other possibilities have been exhausted.  I don't think your attitude is a recipe for improving adoption. 

I have in fact mentioned HexaPDF to another developer within my organization, and he was interested in initially thought the license would be no problem (based on his prior understanding of the AGPL.  But the more he read about it the more confusion differing sources created regarding the license, and he decided to go with Prawn.  I'm sure this scenario is replicated thousands of times.

**Comment by gettalong** [comment]
@brandondrew I know what the license does and when it can be used. *However*, if you are following general news about software licensed under the AGPL (or any similar license) you will find many instances where only lawyers (or judges) were able to decide whether a certain usage pattern was actually allowed by the license.

If you are not sure if **you can use the software under the AGPL license in *your* use case**, consult a lawyer or buy the commercial license. I don't know what you build, how you build it and who your users are, so I can't give you a definite answer.

If you are building the software as part of your job, someone at your company is certainly responsible for making sure that all used licenses are respected. Consult them - they have to know the answer! That is their job!

The choice of license was deliberate and I know that some people will shun the project because of it. But even if this is the case, it nonetheless was and is the best choice.

Prawn is a great PDF creation library and if it fits your needs, great!

**Comment by brandondrew** [comment]
I'm not asking highly specific questions about a corner case.  I'm asking for very general guidance about how you understand the AGPL.

Nor am I shunning the project, and I was very explicit about that.  I do not in the slightest bit begrudge your need to earn money.

It appears that you have no intention of providing any clear guidance as to how you understand the AGPL.  That's your choice,  and perhaps necessary to avoid someone trying to twist your words in court to cover something you never intended.  Fine. I'm somewhat sympathetic, but the lack of clarity is without question costing you users.  I would expect that it's a very large proportion of potential users, but that's speculation.  All I can say with certainty is that it's 100% of the 2 cases I know of.

**Comment by gettalong** [comment]
Thank you for being so constructive about this. And I'm sorry that I can't give you a better answer.

If HexaPDF was provided just under the AGPL, adoption would certainly be very low and some use cases would not be possible. But since there also is the commercial license, you always have the choice of using that.

And there are use cases where the AGPL works fine, one being the `hexapdf` binary that comes with HexaPDF itself. Or small Ruby command line scripts. Or when used in an AGPL software.

**Comment by brandondrew** [comment]
I don't know if you appreciate the extent to which developers are expected to deliver features in a timely way, and consulting a lawyer suddenly means _an indefinite amount of time that is entirely outside of our control_ delaying our progress.  If it turns out that it is _possible_ to deliver the same feature without having to consult the lawyer, our judgement in raising the question will itself be brought into question.  

Never in any organization I've worked for have I been told who the appropriate lawyer is, so there is no indication that the organizations are prepared for such questions on a regular basis.  I would first have to find the person to ask the question of.  While there are undoubtedly lawyers working for every sizable organization, I don't think it is remotely safe to assume that even most of them could answer this question in a timely way.   All of that means development is held up, and developer judgement called into question.  When I've seen development held up by legal questions, it takes months for lawyers to answer what appears to be a very simple question.   Because (1) you refuse to give a clear answer about when and where your library can be used, and (2) trying to find the answer inside an enterprise is like diving into quicksand,  I think virtually all Rails developers will choose the library that is politically and legally "safe", whether or not it is technically superior, unless they have no other option.  (Or they will ignore the law because they don't expect to get caught.)  None of this helps you in the slightest.  

Giving a simple answer would help you _and_ the rest of us a great deal.

**Comment by gettalong** [comment]
@brandondrew I understand what you are saying and I get it that this is often a problem. However, this is something I can't solve.

Regarding an answer: Alas, there is *no* simple answer. The best I can give you: The AGPL's copyleft provision comes into effect when code gets distributed. In case of AGPL distribution also means use via a remote network. So if you are developing an application that is used via a network, for example a web service, and that application uses an AGPL library like HexaPDF somewhere, you have to provide the complete source code of your application to each user. If you are fine with that, use the AGPL version. Otherwise buy a commercial license.

**Comment by cedm** [comment]
I feel compelled to comment on this issue as the information provided is factually wrong. The AGPL, like other GNU licences (GPL, LGPL), is a strong copyleft licence. In layman's terms, this means that the mere usage of a xGPL-licensed software is NOT restricted in any way. The provisions and restrictions only applies when the licensed code is both modified AND distributed.

For the AGPL licence, distribution also include remote usage. But the key requirement remains that the licensed code is modified. Using this gem to produce a PDF is not modifying the licensed code, and therefore the license provisions do not apply.

If one were to modify this gem and redistribute it however, then the modifications made to the gem would have to be made available under the same terms as the original code. But in no way this would "taint" a software that merely makes calls to the gem's API.

All this is discussed at length at https://www.gnu.org/licenses/gpl-faq.html

**Comment by cedm** [comment]
To the author of this gem, and this was your intention, you should make that abundantly clear at the top of your README.md file. A lot of people are going to be caught off guard otherwise.

**Comment by gettalong** [comment]
@cedm Yes, I'm aware of this and yes, this is intentional. Otherwise it would not be possible to sell commercial licenses. However, there are many situation where AGPL HexaPDF can be used. E.g. if you just use it personally, even for commercial stuff, AGPL works fine, since there is no distribution. Or if you post-process PDFs using the `hexapdf` CLI.

If HexaPDF was LGPL (or where it is even clearer the MIT) it could be used in a SaaS without a commercial license. Since this is one of the main use cases for a PDF library nowadays, I decided to use AGPL+commercial license to sustain the development of HexaPDF.

The other possible licensing solution was open core plus paid add-ons (e.g. like sidekiq) but I don't think this is a good fit for me and HexaPDF.

I will add a statement about licensing at the top to make this clearer.

**Comment by brandondrew** [comment]
> If one were to modify this gem and redistribute it however, then the modifications made to the gem would have to be made available under the same terms as the original code. ~But in no way this would "taint" a software that merely makes calls to the gem's API.~

You were really getting my hopes up for a little while...

> EDIT: I was wrong regarding my last statement. Gems used in an application are considered a linked modules (
...
> So legally, it's pay for the commercial license or open source your app (or go rogue I guess?).

until you came to the same conclusion that I feared was correct.

It requires a **_very bizarre_** definition of "distribute code" and an equally strange understanding of "modify", but that's where we seem to be.

(My comments on the language of the AGPL are **_not in any way_** meant as criticisms of Thomas or his work, or of his choice of license.  He is **_absolutely entitled to expect payment_** for his hard work.  It is only a criticism of the wording of the license itself.  Nothing more or less.)

**Comment by gettalong** [comment]
Just for posterity's sake here is an answer on StackOverflow for a very similar topic which might help a bit: https://opensource.stackexchange.com/a/5005

The license scheme of HexaPDF is more or less the same as iText's licensing scheme.
